cpu-checker/check-bios-nx.1
.TH check-bios-nx 1 "9 March 2010" cpu-checker "cpu-checker"
.SH NAME
\fBcheck\-bios\-nx\fP \- determine if BIOS has blocked CPU's NX capabilities

.SH SYNOPSIS
\fBcheck\-bios\-nx [\-h|\-\-help] [\-\-verbose]\fP

.SH OPTIONS
\fB\-h|\-\-help\fP - display a help message and exit

\fB\-\-verbose\fP - explain in detail what has been detected

.SH DESCRIPTION
This program attempts to determine if the running x86-based CPU has NX capabilities (regardless of it being disabled by the BIOS).

If the CPU is NX-capable but the nx bit is missing from flags, exit 1 (i.e. "BIOS settings need changing"), otherwise exit 0 (i.e. "nothing wrong with BIOS")

.SH MORE INFORMATION
\fIhttp://launchpad.net/cpu-checker\fP
.PD

.SH AUTHOR
This manpage was written by Dustin Kirkland <kirkland@canonical.com> for Ubuntu systems (but may be used by others).  Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 3 published by the Free Software Foundation.

On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common-licenses/GPL-3.


vmbuilder/NEWS
0.12.5
  * Ported debian so that users can also runb debian-vm-builder

0.12.4:

  * Revive firstscripts plugins. This fixes the --firstboot and --firstlogin
    options, both of which went missing with 0.12.
  * Make Plugin.install_file work in non-context plugins (plugins that are not
    Distros or Hypervisors).
  * Do a slightly better job at cleaning up after ourselves by removing the
    distro chroot by default and removing the temporary root mount point.
  * Revive QEMu hypervisor.

0.12.3:

 * Reworked the way timezones and locales are handled. This fixes an
   actual problem with /etc/localtime not being set correctly, and fixes
   a theoretical problem with people not running VMBuilder on Ubuntu.
 * Resurrect -o (--overwrite) which went missing with 0.12.0.
 * Resurrect Xen support which went missing with 0.12.0.

0.12.2:

 * Moved test directory under VMBuilder/ and include it in the tarball.
 * Revive ubuntu-vm-builder (which rotted during the 0.11 -> 0.12 transition)
 * Skip tests that need root, if not run as root.
 * Fix installing onto logical volumes.
 * Fix --raw handling.
 * Refresh documenation in disk.py
 * Improve handling of non-string settings from command line and config files.
 * A big bunch of cleanups, most thanks to pyflakes.

0.12.1:

 * Added NEWS file
 * IntegerSetting.set_value now attempts to convert the given value
   to an int instead of rejecting it outright.
 * Bypass blkid cache.
 * Move initctl out of the way during package installs in Ubuntu plugin.
 * Update VMWare driver to 0.12 API
 * Support separate /boot partitions.
 * Update --part handling to 0.12 API.
 * Make sure the given MAC is used in the libvirt domain template.
 * Update libvirt plugin to 0.12 API.
 * Check for hvm capabilities during preflight check.
 * Add option to specify a seedfile which will be fed into the debconf
   database. This enables installing packages like sun-java6-jre which
   requires the user to accept an EULA.
 * Correctly handle config keys with dashes in them.
 * Update postinst plugin to 0.12 API.
 * If trying to install a file to a nonexisting directory, the directory
   will be created first.
 * Refresh the valid_flavours for jaunty, karmic, and Lucid.
 * Remove lpia from Lucid.
 * Rewrite $LANG (s/utf8$/UTF-8/) so that locale-gen is happy again.
 * Bail out early if destdir already exists.
 * Re-add the option to specify a custom template directory.
 * Re-add the option to specify a config file.
 * Default to ext3 for Dapper through Jaunty
 * Support non-dhcp network configurations.



cpu-checker/old/check-bios-nx.sh
#!/bin/sh
# Copyright (C) 2009-2010 Canonical, Ltd.
# License: GPLv3
# Author: Kees Cook <kees@ubuntu.com>
#
# Attempts to determine if the running x86-based CPU has NX capapbilities
# (regardless of it being disabled by the BIOS).  If the CPU is NX-capable
# but the nx bit is missing from flags, exit 1 (i.e. "BIOS settings need
# changing"), otherwise exit 0 (i.e. "nothing wrong with BIOS")
#
# lacks NX:
#     not pae:
#         cpu family <= 5
#         cpu family > 6 && cpu family < 15
#         cpu family == 6, model <= 12
#     pae, cpu family == 6, model == 13  (excepting some sSpec?)
#             http://processorfinder.intel.com/List.aspx?ParentRadio=All&ProcFam=942&SearchKey=
# has NX:
#     http://processorfinder.intel.com/Default.aspx
#     pae, cpu family == 6, model >= 14
#     pae, cpu family == 15, model >= 3
#     pae, cpu family > 15
set -e
export LANG=C

usage() {
    echo "Usage: $0 [options]"
    echo ""
    echo "Options:"
    echo "  -h, --help  show this help message and exit"
    echo "  --verbose   Explain in detail what has been detected"
}

VERBOSE=
verbose() {
    if [ -n "$VERBOSE" ]; then
        echo "$@" >&2
    fi
}

TEMP=$(getopt -o h --long verbose,help -n check-bios-nx -- "$@")
eval set -- "$TEMP"

while :; do
    case "$1" in
        -h|--help) usage ; exit 0 ;;
        --verbose) VERBOSE=1; shift ;;
        --) shift ; break ;;
        *) usage >&2 ; exit 2 ;;
    esac
done

ARCH=${CHECK_BIOS_NX_MACHINE:-$(uname -m)}
if ! echo "$ARCH" | egrep -q '^(i.86|x86_64)$' ; then
    verbose "This script is currently only useful on x86-based CPUs"
    exit 0
fi

# Avoid calling grep/head/awk 3 times in a row, thanks to Jamie Strandboge
# and Steve Beattie.
set -- $(awk 'BEGIN { FS=": " ; family = "0" ; model = "0" ; flags = "-" } \
     family == "0" && /^cpu family\t/ { family = $2 } \
     model  == "0" && /^model\t/      { model = $2 } \
     flags  == "-" && /^flags\t/      { flags = $2 } \
     END { print family; print model; print flags }' \
    ${CHECK_BIOS_NX_CPUINFO:-/proc/cpuinfo})
family=$1
model=$2
shift 2
flags=$*

if [ -z "$flags" ]; then
    # No flags found (?!), fail open
    verbose "No 'flags' were found for this CPU.  Check /proc/cpuinfo"
    exit 1
fi

# If it's in the flags, it's not being disabled by the BIOS; rejoice.
if echo " $flags " | grep -q ' nx ' ; then
    verbose "This CPU has 'nx' in the flags, so the BIOS is not disabling it."
    exit 0
fi

if echo " $flags " | grep -q ' pae ' ; then
    if [ -z "$model" ] || [ -z "$family" ]; then
        # Cannot identify CPU, fail open
        verbose "No 'model' or 'family' were found for this CPU.  Check /proc/cpuinfo"
        exit 1
    fi

    if ([ $family -eq 6 ] && [ $model -ge 14 ]) || \
       ([ $family -eq 15 ] && [ $model -ge 3]) || \
       [ $family -gt 15 ]; then
        # NX should be available in CPU, but missing from flags
        echo "This CPU is family $family, model $model, and has NX capabilities but is unable to" >&2
        echo "use these protective features because the BIOS is configured to disable" >&2
        echo "the capability.  Please enable this in your BIOS.  For more details, see:" >&2
        echo "https://wiki.ubuntu.com/Security/CPUFeatures" >&2
        exit 1
    else
        # NX not available in CPU
        verbose "This CPU is family $family, model $model, and does not have NX capabilities."
        exit 0
    fi

fi

verbose "This CPU is not PAE capable, so it does not have NX."
exit 0
